src/databricks_ai_bridge/lakebase.py

```diff
@@ -0,0 +1,352 @@
+        return super().connect(conninfo, **kwargs)
+
+
+def _make_rotating_connection_class(
+    workspace_client: WorkspaceClient, instance_name: str, cache_duration_sec: int
+) -> type[RotatingCredentialConnection]:
+    return type(
+        f"LakebaseRotatingConnection_{instance_name}",
+        (RotatingCredentialConnection,),
+        {
+            "workspace_client": workspace_client,
+            "instance_name": instance_name,
+            "cache_duration_sec": cache_duration_sec,
+            "_cache_lock": Lock(),
+            "_cached_token": None,
+            "_cache_ts": None,
+        },
+    )
+
+
+def _infer_username(w: WorkspaceClient) -> str:
+    """Resolve a default username preferring service-principal identity."""
+    try:
+        sp = w.current_service_principal.me()
+        if sp and getattr(sp, "application_id", None):
+            return sp.application_id
+    except Exception:
+        logger.debug(
+            "Could not get service principal, using current user for Lakebase credentials."
+        )
+
+    user = w.current_user.me()
+    return user.user_name
+
+
+class LakebasePool:
+    """Wrapper around a psycopg connection pool with rotating credentials."""
+
+    def __init__(
+        self,
+        *,
+        instance_name: str,
+        workspace_client: WorkspaceClient | None = None,
+        host: str | None = None,
+        database: str | None = None,
+        username: Optional[str] = None,
+        port: Optional[int] = None,
+        sslmode: Optional[str] = None,
+        token_cache_minutes: Optional[int] = None,
+        connection_kwargs: Optional[dict[str, object]] = None,
+        **pool_kwargs: object,
+    ) -> None:
+        if workspace_client is None:
+            workspace_client = WorkspaceClient()
+
+        if host is None:
+            host = DEFAULT_HOST
+        if host is None:
+            raise ValueError(
+                "Lakebase host must be provided. Specify the host argument or set the DB_HOST environment variable."
+            )
+
+        if database is None:
+            database = DEFAULT_DATABASE
+        if port is None:
+            port = DEFAULT_PORT
+        if sslmode is None:
+            sslmode = DEFAULT_SSLMODE
+        cache_minutes = (
+            DEFAULT_CACHE_MINUTES if token_cache_minutes is None else int(token_cache_minutes)
+        )
+
+        pool_kwargs = dict(pool_kwargs)
+        for reserved in ("conninfo", "connection_class", "kwargs"):
+            if reserved in pool_kwargs:
+                raise TypeError(f"Argument '{reserved}' cannot be overridden.")
+        min_size = int(pool_kwargs.pop("min_size", DEFAULT_MIN_SIZE))
+        max_size = int(pool_kwargs.pop("max_size", DEFAULT_MAX_SIZE))
+        timeout = float(pool_kwargs.pop("timeout", DEFAULT_TIMEOUT))
+        open_flag = pool_kwargs.pop("open", True)
+
+        self.workspace_client = workspace_client
+        self.instance_name = instance_name
+        self.host = host
```

Comment:
Is this ever used (outside of logger statements)?

Reply:
Oh I see it's used in `conninfo`, but I think the host can be derived from the lakebase instance name, no need to pass in an override

Reply:
The host is separate from the lakebase name, example instance name is just anything you give it like "lakebase-name" but host is built from the instance uuid, something like `instance-cd864d21-d19f-4180-bddd-eb1bf4da7625.database.staging.cloud.databricks.com`. Are you thinking of using another method to derive the host from the lakebase instance name here?

Reply:
just confirmed the host is used in the connection string, but the lakebase instance name is not. The connection string is composed like:
`psql "host=instance-cd864d21-d19f-4180-bddd-eb1bf4da7625.database.staging.cloud.databricks.com user=jenny.sun@databricks.com dbname=databricks_postgres port=5432 sslmode=require"` 

In the current CUJ, the lakebase name is passed in the config defined by the user for the purposes of adding the lakebase resource to specify for automatic auth passthrough at deployment time. So instead of mandating passing in instance name, we'd have to do so for host name